File: weiqi-ai/core/board.py
"""
围棋棋盘模块

实现棋盘的基本操作，包括落子、提子、自杀禁令等。
必须与 Go 后端的实现保持 100% 一致。
"""

from enum import IntEnum
from typing import List, Tuple, Set
from dataclasses import dataclass
import hashlib


# 棋盘大小
BOARD_SIZE = 19


class Player(IntEnum):
    """玩家/棋子类型"""
    EMPTY = 0
    BLACK = 1
    WHITE = 2


@dataclass
class Point:
    """棋盘上的一个坐标点"""
    x: int
    y: int

    def __hash__(self):
        return hash((self.x, self.y))

    def __eq__(self, other):
        if not isinstance(other, Point):
            return False
        return self.x == other.x and self.y == other.y


class BoardError(Exception):
    """棋盘操作错误的基类"""
    pass


class PointOutOfBoundsError(BoardError):
    """坐标超出棋盘范围"""
    pass


class PointNotEmptyError(BoardError):
    """该位置已有棋子"""
    pass


class SuicideMoveError(BoardError):
    """自杀手（非法）"""
    pass


class Board:
    """
    围棋棋盘
    
    使用 19x19 的二维数组表示棋盘状态。
    坐标系：左上角为 (0, 0)，右下角为 (18, 18)
    """

    def __init__(self):
        """创建一个空棋盘"""
        self.grid: List[List[Player]] = [
            [Player.EMPTY for _ in range(BOARD_SIZE)]
            for _ in range(BOARD_SIZE)
        ]

    def place_stone(self, player: Player, point: Point) -> int:
        """
        在指定位置落子
        
        Args:
            player: 落子的玩家（黑或白）
            point: 落子位置
            
        Returns:
            提掉的对方棋子数量
            
        Raises:
            PointOutOfBoundsError: 坐标超出范围
            PointNotEmptyError: 该位置已有棋子
            SuicideMoveError: 自杀手（非法）
        """
        # 1. 基础合法性检查
        if not self._is_on_board(point):
            raise PointOutOfBoundsError(f"Point ({point.x}, {point.y}) is outside the board")
        
        # 注意：grid[行][列]，而 Point.x 是列，Point.y 是行
        if self.grid[point.y][point.x] != Player.EMPTY:
            raise PointNotEmptyError(f"Point ({point.x}, {point.y}) is not empty")

        # 2. 试探性落子
        self.grid[point.y][point.x] = player

        # 3. 检查并移除对方被提的子
        opponent = self._get_opponent(player)
        captured_stones: List[Point] = []
        
        for neighbor in self._get_neighbors(point):
            if self.grid[neighbor.y][neighbor.x] == opponent:
                group, liberties = self._find_group_and_liberties(neighbor)
                if liberties == 0:
                    captured_stones.extend(group)
                    for stone in group:
                        self.grid[stone.y][stone.x] = Player.EMPTY

        # 4. 自杀禁令检查
        _, new_liberties = self._find_group_and_liberties(point)
        if new_liberties == 0:
            # 这是自杀手，回滚所有操作
            self.grid[point.y][point.x] = Player.EMPTY
            for stone in captured_stones:
                self.grid[stone.y][stone.x] = opponent
            raise SuicideMoveError("Suicide move is not allowed")

        return len(captured_stones)

    def _is_on_board(self, point: Point) -> bool:
        """检查坐标是否在棋盘范围内"""
        return 0 <= point.x < BOARD_SIZE and 0 <= point.y < BOARD_SIZE

    def _get_neighbors(self, point: Point) -> List[Point]:
        """获取一个点的所有合法邻居（上下左右）"""
        neighbors = []
        directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
        
        for dx, dy in directions:
            x, y = point.x + dx, point.y + dy
            if 0 <= x < BOARD_SIZE and 0 <= y < BOARD_SIZE:
                neighbors.append(Point(x, y))
        
        return neighbors

    def _get_opponent(self, player: Player) -> Player:
        """获取对手的颜色"""
        if player == Player.BLACK:
            return Player.WHITE
        return Player.BLACK

    def _find_group_and_liberties(self, start_point: Point) -> Tuple[List[Point], int]:
        """
        使用 BFS 寻找一个点所在的棋块及其气数
        
        Args:
            start_point: 起始点
            
        Returns:
            (棋块中的所有点, 气的数量)
        """
        if self.grid[start_point.y][start_point.x] == Player.EMPTY:
            return [], 0

        player = self.grid[start_point.y][start_point.x]
        group: List[Point] = []
        visited_in_group: Set[Point] = {start_point}
        visited_liberties: Set[Point] = set()
        queue: List[Point] = [start_point]

        while queue:
            current = queue.pop(0)
            group.append(current)

            for neighbor in self._get_neighbors(current):
                neighbor_value = self.grid[neighbor.y][neighbor.x]
                
                if neighbor_value == Player.EMPTY:
                    visited_liberties.add(neighbor)
                elif neighbor_value == player and neighbor not in visited_in_group:
                    visited_in_group.add(neighbor)
                    queue.append(neighbor)

        return group, len(visited_liberties)

    def state_hash(self) -> str:
        """
        生成当前棋盘状态的唯一哈希值
        
        用于实现全局同形再现禁令（Ko 规则）
        """
        state_str = ''.join(
            str(int(self.grid[i][j]))
            for i in range(BOARD_SIZE)
            for j in range(BOARD_SIZE)
        )
        return hashlib.md5(state_str.encode()).hexdigest()

    def clone(self) -> 'Board':
        """创建棋盘的深拷贝"""
        new_board = Board()
        new_board.grid = [row[:] for row in self.grid]
        return new_board

    def to_list(self) -> List[List[int]]:
        """将棋盘转换为二维列表（用于 JSON 序列化）"""
        return [[int(self.grid[i][j]) for j in range(BOARD_SIZE)] for i in range(BOARD_SIZE)]

    @classmethod
    def from_list(cls, grid_list: List[List[int]]) -> 'Board':
        """从二维列表创建棋盘"""
        board = cls()
        for i in range(BOARD_SIZE):
            for j in range(BOARD_SIZE):
                board.grid[i][j] = Player(grid_list[i][j])
        return board

    def __str__(self) -> str:
        """返回棋盘的字符串表示（用于调试）"""
        lines = ["   " + " ".join(f"{i:2d}" for i in range(BOARD_SIZE))]
        
        for i in range(BOARD_SIZE):
            row = f"{i:2d} "
            for j in range(BOARD_SIZE):
                if self.grid[i][j] == Player.EMPTY:
                    row += " . "
                elif self.grid[i][j] == Player.BLACK:
                    row += " X "
                else:
                    row += " O "
            lines.append(row)
        
        return "\n".join(lines)

File: game/board.go
package game

import (
	"encoding/json"
	"errors"
	"fmt"
	"strings"
)

// BoardSize 定义了棋盘的大小
const BoardSize = 19

// Player 定义了棋盘上的点的状态: 空, 黑棋, 白棋
type Player int8

const (
	Empty Player = 0
	Black Player = 1
	White Player = 2
)

// MarshalJSON 自定义 JSON 序列化
func (p Player) MarshalJSON() ([]byte, error) {
	switch p {
	case Black:
		return []byte(`"Black"`), nil
	case White:
		return []byte(`"White"`), nil
	default:
		return []byte(`"Empty"`), nil
	}
}

// UnmarshalJSON 自定义 JSON 反序列化
func (p *Player) UnmarshalJSON(data []byte) error {
	str := string(data)
	switch str {
	case `"Black"`, `"black"`, `"1"`:
		*p = Black
	case `"White"`, `"white"`, `"2"`:
		*p = White
	default:
		*p = Empty
	}
	return nil
}

// Point 定义了棋盘上的一个坐标
type Point struct {
	X int `json:"x"`
	Y int `json:"y"`
}

// Board 定义了棋盘的结构
type Board struct {
	Grid [BoardSize][BoardSize]Player `json:"-"`
}

// MarshalJSON 自定义 Board 的 JSON 序列化，将 Grid 转换为数字数组
func (b *Board) MarshalJSON() ([]byte, error) {
	// 创建数字数组
	grid := make([][]int8, BoardSize)
	for i := 0; i < BoardSize; i++ {
		grid[i] = make([]int8, BoardSize)
		for j := 0; j < BoardSize; j++ {
			grid[i][j] = int8(b.Grid[i][j])
		}
	}
	
	// 序列化为 JSON
	type Alias struct {
		Grid [][]int8 `json:"grid"`
	}
	return json.Marshal(&Alias{
		Grid: grid,
	})
}

// UnmarshalJSON 自定义 Board 的 JSON 反序列化
func (b *Board) UnmarshalJSON(data []byte) error {
	type Alias struct {
		Grid [][]int8 `json:"grid"`
	}
	var alias Alias
	if err := json.Unmarshal(data, &alias); err != nil {
		return err
	}
	
	// 转换回 Player 类型
	for i := 0; i < BoardSize && i < len(alias.Grid); i++ {
		for j := 0; j < BoardSize && j < len(alias.Grid[i]); j++ {
			b.Grid[i][j] = Player(alias.Grid[i][j])
		}
	}
	
	return nil
}

// 定义一些常亮错误, 便于调用方法判断
var (
	ErrPointOutOfBounds = errors.New("point is outside the board")
	ErrPointNotEmpty    = errors.New("point is not empty")
	ErrSuicideMove      = errors.New("suicide move is not allowed")
)

// NewBoard 创建并返回一个空的棋盘
func NewBoard() *Board {
	return &Board{}
}

// PlaceStone 在指定坐标落下指定颜色的旗子
// 包含了对落子合法性的基础检查
// 它会处理提子逻辑，并返回提子的数量
func (b *Board) PlaceStone(player Player, p Point) (captures int, err error) {
	// 1. 基础合法性检查
	// 检查坐标是否在棋盘内
	if p.X < 0 || p.X >= BoardSize || p.Y < 0 || p.Y >= BoardSize {
		return 0, ErrPointOutOfBounds
	}
	// 检查该位置是否为空
	// 注意：Grid[行][列]，而 Point.X 是列，Point.Y 是行
	if b.Grid[p.Y][p.X] != Empty {
		return 0, ErrPointNotEmpty
	}

	// 2. 试探性地落子
	b.Grid[p.Y][p.X] = player

	// 3. 检查并移除对方被提的子
	opponent := getOpponent(player)
	var capturedStones []Point
	for _, n := range getNeighbors(p) {
		if b.Grid[n.Y][n.X] == opponent {
			group, liberties := b.findGroupAndLiberties(n)
			if liberties == 0 {
				capturedStones = append(capturedStones, group...)
				for _, stone := range group {
					b.Grid[stone.Y][stone.X] = Empty
				}
			}
		}
	}

	// 4. 自杀禁令检查
	// 在提子后，检查落子点所在的棋块是否还有气
	_, newLiberties := b.findGroupAndLiberties(p)
	if newLiberties == 0 {
		// 这是一个自杀点，回滚所有操作
		b.Grid[p.Y][p.X] = Empty               // 撤销落子
		for _, stone := range capturedStones { // 将被提的子放回去
			b.Grid[stone.Y][stone.X] = opponent
		}
		return 0, ErrSuicideMove
	}

	return len(capturedStones), nil
}

// findGroupAndLiberties 使用广度优先搜索(BFS)寻找一个点所在的棋块及其气数
// visitedInGroup 用于记录棋块内的点，防止重复搜索
// visitedLiberties 用于记录气点，防止重复计数
func (b *Board) findGroupAndLiberties(startPoint Point) (group []Point, liberties int) {
	if b.Grid[startPoint.Y][startPoint.X] == Empty {
		return nil, 0
	}

	player := b.Grid[startPoint.Y][startPoint.X]
	q := []Point{startPoint}
	visitedInGroup := map[Point]bool{startPoint: true}
	visitedLiberties := map[Point]bool{}

	for len(q) > 0 {
		current := q[0]
		q = q[1:]
		group = append(group, current)

		for _, n := range getNeighbors(current) {
			switch b.Grid[n.Y][n.X] {
			case Empty:
				if !visitedLiberties[n] {
					visitedLiberties[n] = true
				}
			case player:
				if !visitedInGroup[n] {
					visitedInGroup[n] = true
					q = append(q, n)
				}
			}
		}
	}

	return group, len(visitedLiberties)
}

// getNeighbors 返回一个点的所有合法邻居坐标
func getNeighbors(p Point) []Point {
	neighbors := make([]Point, 0, 4)
	if p.X > 0 {
		neighbors = append(neighbors, Point{X: p.X - 1, Y: p.Y})
	}
	if p.X < BoardSize-1 {
		neighbors = append(neighbors, Point{X: p.X + 1, Y: p.Y})
	}
	if p.Y > 0 {
		neighbors = append(neighbors, Point{X: p.X, Y: p.Y - 1})
	}
	if p.Y < BoardSize-1 {
		neighbors = append(neighbors, Point{X: p.X, Y: p.Y + 1})
	}
	return neighbors
}

// getOpponent 返回对手的颜色
func getOpponent(player Player) Player {
	if player == Black {
		return White
	}
	return Black
}

// StateHash 将当前棋盘状态转换为一个唯一的字符串，用作哈希键
func (b *Board) StateHash() string {
	var sb strings.Builder
	sb.Grow(BoardSize * BoardSize) // 预分配内存以提高性能
	for i := 0; i < BoardSize; i++ {
		for j := 0; j < BoardSize; j++ {
			sb.WriteByte(byte(b.Grid[i][j]) + '0') // 将 0,1,2 转换为 '0','1','2'
		}
	}
	return sb.String()
}

// Clone 创建并返回当前棋盘的一个深拷贝
func (b *Board) Clone() *Board {
	clone := NewBoard()
	clone.Grid = b.Grid // 在 Go 中，数组是值类型，直接赋值就是拷贝
	return clone
}

// String 方法让 Board 类型可以被方便地被打印出来, 用于调试
func (b *Board) String() string {
	var sb strings.Builder
	sb.WriteString("   ")
	for i := 0; i < BoardSize; i++ {
		sb.WriteString(fmt.Sprintf("%2d ", i))
	}
	sb.WriteString("\n")

	for i := 0; i < BoardSize; i++ {
		sb.WriteString(fmt.Sprintf("%2d ", i))
		for j := 0; j < BoardSize; j++ {
			switch b.Grid[i][j] {
			case Empty:
				sb.WriteString(" . ")
			case Black:
				sb.WriteString(" X ")
			case White:
				sb.WriteString(" O ")
			}
		}
		sb.WriteString("\n")
	}
	return sb.String()
}
